R/rt_login.R

```diff
@@ -2,59 +2,62 @@
 #'
 #' Use this to log into RT at the start of your session.
 #'
-#' @param base (character) The base URL that hosts RT for your organization
 #' @param user (character) Your username
 #' @param pass (character) Your password
+#' @param rt_base (character) The base URL that hosts RT for your organization. Set the base URL in your R session using \code{options(rt_base = "https://server.name/rt/")}
 #'
 #' @export
-#' 
+#'
 #' @import httr
 #' @import stringr
 #'
 #' @examples
 #' \dontrun{
-#' rt_login("https://server.name/rt/", "my_username", "my_password")
+#' options(rt_base = "https://server.name/rt/")
+#' rt_login("my_username", "my_password")
 #' }
 
-rt_login <- function(base, user, pass) {
-  req <- httr::POST(rt_url(base), body = list('user' = user, 'pass' = pass))
-
-  # Process RT's strange custom status code
-  status_match <- stringr::str_match_all(httr::content(req), "RT\\/[\\d\\.]+ (\\d+).+")
-
-  if (length(status_match) != 1 && all(dim(status_match[[1]]) != c(1, 2))) {
-    stop(call. = FALSE, "Failed to parse response from RT.")
+rt_login <- function(user, pass, rt_base = getOption("rt_base")) {
+  if(!is.character(rt_base)){
+    stop('Check your base URL. Set it in your R session using option(rt_base = "https://server.name/rt/")', call. = FALSE)
   }
 
-  status_code <- as.numeric(status_match[[1]][,2])
+  base_api <- paste(stringr::str_replace(rt_base, "\\/$", ""), # removes trailing slash from base URL just in case
+                "REST", "1.0", sep = "/")
+  req <- httr::POST(base_api, body = list('user' = user, 'pass' = pass))
 
-  if (status_code == 200) {
-    message("Successfully logged in.")
+  # Check that login worked
+  # Try getting a ticket
+  test_ticket <- httr::GET(paste(base_api, "ticket", 00000, sep = "/"))
```

Comment:
Good idea! I think you can actually test whether login failed without an additional request though. The origin `POST` to log in returns a status of 200 (which is wrong) but with a response body that contains the message I see you're checking for after you making you second call (the `GET)`:

```r
r$> httr::POST('https://support.nceas.ucsb.edu/rt/REST/1.0', body = list('user' = 'asdf', 'pass' = 'asd'))
Response [https://support.nceas.ucsb.edu/rt/REST/1.0]
  Date: 2018-04-24 21:53
  Status: 200
  Content-Type: text/plain; charset=utf-8
  Size: 74 B
RT/4.4.1 401 Credentials required

Your username or password is incorrect

r$> content(.Last.value)
[1] "RT/4.4.1 401 Credentials required\n\nYour username or password is incorrect\n"
```

In an ideal world, RT would return the login request with a status code in the 400 range.